tests/lib/thrift/ThriftTest/src/client.cpp

```diff
@@ -0,0 +1,509 @@
+  truck2.i32_thing = 2;
+  truck2.i64_thing = 2;
+  insane.xtructs.push_back(truck);
+  insane.xtructs.push_back(truck2);
+
+  map<UserId, map<Numberz::type, Insanity>> whoa;
+  testClient.testInsanity(whoa, insane);
+
+  zassert_equal(whoa.size(), 2, NULL);
+
+  map<UserId, map<Numberz::type, Insanity>>::const_iterator it1 = whoa.find(UserId(1));
+  zassert_not_equal(it1, whoa.end(), NULL);
+
+  if (it1 != whoa.end()) {
+    auto it12 = it1->second.find(Numberz::TWO);
+    zassert_not_equal(it12, it1->second.end(), NULL);
+    zassert_equal(it12->second, insane, NULL);
+
+    auto it13 = it1->second.find(Numberz::THREE);
+    zassert_not_equal(it13, it1->second.end(), NULL);
+    zassert_equal(it13->second, insane, NULL);
+  }
+
+  map<UserId, map<Numberz::type, Insanity>>::const_iterator it2 = whoa.find(UserId(2));
+  zassert_not_equal(it2, whoa.end(), NULL);
+  if (it2 != whoa.end()) {
+    auto it26 = it2->second.find(Numberz::SIX);
+    zassert_not_equal(it26, it2->second.end(), NULL);
+    zassert_equal(it26->second, Insanity(), NULL);
+  }
+}
+
+static void exception_test(void) {
+  ThriftTestClient& testClient = *(context.client);
+  try {
+    testClient.testException("Xception");
+    zassert_unreachable("Xception: Unhandled exception should abort execution");
+  } catch (Xception& e) {
+  }
+
+  try {
+    testClient.testException("TException");
+    zassert_unreachable("TException: Unhandled exception should abort execution");
+  } catch (const TException&) {
+  }
+
+  try {
+    testClient.testException("success");
+  } catch (exception& ex) {
+    zassert_exception("%s", ex.what());
+  }
+}
+
+static void multi_exception_test(void) {
+  ThriftTestClient& testClient = *(context.client);
+  try {
+    Xtruct result;
+    testClient.testMultiException(result, "Xception", "test 1");
+    zassert_unreachable("Xception: Unhandled exception should abort execution");
+  } catch (Xception& e) {
+  }
+
+  try {
+    Xtruct result;
+    testClient.testMultiException(result, "Xception2", "test 2");
+    zassert_unreachable("Xception2: Unhandled exception should abort execution");
+
+  } catch (Xception2& e) {
+  }
+
+  try {
+    Xtruct result;
+  } catch (exception& ex) {
+    zassert_exception("%s", ex.what());
+  }
+}
+
+static void oneway_void_test(void) {
+  ThriftTestClient& testClient = *(context.client);
+  uint64_t startOneway = now();
+  testClient.testOneway(1);
+  uint64_t elapsed = now() - startOneway;
+  zassert_true(elapsed <= 200 * 1000 /* 0.2 seconds */, "took %.2f ms", (double)elapsed / 1000.0);
+}
+
+static shared_ptr<ThriftTestClient> setup_client(const string& my_addr, const int port) {
+  std::shared_ptr<TTransport> trans(new TSocket(my_addr, port));
+  std::shared_ptr<TTransport> transport(new TBufferedTransport(trans));
+  std::shared_ptr<TProtocol> protocol(new TBinaryProtocol(transport));
+  transport->open();
+  return std::make_shared<ThriftTestClient>(protocol);
+}
+
+static void* server_func(void* arg) {
```

Comment:
`server_func`, `setup`, `teardown`, `setup_client`, and `setup_server` can go into `main.cpp`.

I guess the main thing that needs to go into `TestUtils.hpp` is the `struct context` definition & declaration. Maybe it's best to name it `context.hpp`

Reply:
`setup_server` need a complete declaration (rather than forward declaration) of `class TestHandler`, whose declaration and definition were put together in `server.cpp`. If `setup_server` has to be moved to `main.cpp`, a header contains the complete declaration of `class TestHandler` is required. I'll just rename `server.cpp` to `server.hpp` and include it in `main.cpp` to do this for now.